Fixed the authentication done when using build to a private repository API version 1.20

walbus opened the pull request:
AuthRegistryConfig looked like:

{"configs":{"myrepository":{
     "username":"myusername",
     "password":"mypassword",
     "auth":"",
     "email":"myemail",
     "serveraddress":"myserverAddress"}}}

But needs to look like:
{"myrepository":{
     "username":"myusername",
     "password":"mypassword",
     "auth":"",
     "email":"myemail",
     "serveraddress":"myserverAddress"}}

Added the fix to AuthRegistryConfig.
Added a test to make sure the test works

walbus commented:
This fix is for Docker API version 1.20

davidxia commented:
We're still using docker API 1.18. Is there a way you can make this backwards compatible?

walbus commented:
Sure. 
I will see if I can make a nice solution that will make it easy to take future changes into account.
Somehow I have to check for the version and then decide depending on that.

walbus commented:
I ran tests on a machine using api-version 1.18, 1.19, 1.20, and 1.21. It should work now.

davidxia commented:
@walbus Thanks. Can you cleanup the git history a bit to make it easier for us to review?

walbus commented:
@davidxia I tried, but git is not working as I expected it too.
That is definitely my fault, since I am new to git. I tried to look at different sources about how to fix the history in git, but it keeps doing strange things when I try to rebase or merge.

walbus commented:
@davidxia What I tried to do here is: 
I first made a branch of your master to fix the authentication.
I then added code to make it backwards compatible.
Since your master had moved forward. I merged from your master.
Somehow this was understood by git as being my own new commits?
So now it thinks I am still behind your master, while also being in front of the master with commit duplicates from the master?

I have no idea why I can't just make git understand I only made changes to three files and added a third.

I removed the Jackson JSON annotations from AuthRegistryConfig.java and added 
    
    @JsonAnyGetter
    public Map<String, Map<String, String>> getConfigs() {
        return configs;
    }

In DefaultDockerClient.java authRegistryHeader method I added
   
    String apiVersion = version().apiVersion();
    int versionComparison = versionCompare(apiVersion, "1.19");

    // Version below 1.19
    if (versionComparison < 0) {
        authRegistryJson = "{\"configs\":" + authRegistryJson + "}";
    }
    // Version equal 1.19
    else if (versionComparison == 0) {
        authRegistryJson = "{\"auths\":" + authRegistryJson + "}";
    }

In DefaultDockerClientTest,java I added a test:

    @Test
    public void testBuildPrivateRepoWithAuth() throws Exception {
    final String dockerDirectory = Resources.getResource("dockerDirectoryNeedsAuth").getPath();
    AuthConfig authConfig = AuthConfig.builder().email(AUTH_EMAIL2).username(AUTH_USERNAME2)
            .password(AUTH_PASSWORD2).build();

    final DefaultDockerClient sut2 = DefaultDockerClient.builder()
            .uri(dockerEndpoint)
            .authConfig(authConfig)
            .build();

    sut2.build(Paths.get(dockerDirectory),
            "testauth",
            DockerClient.BuildParameter.PULL_NEWER_IMAGE);
    }

Lastly I created a Dockerfile in a new folder resources/dockerDirectoryNeedsAuth/Dockerfile that contains:

    FROM dxia2/scratch-private:latest

But for some reason I can't get git to not have conflicts that I have to solve multiple times...

walbus commented:
@davidxia Now it should look right.

davidxia commented:
@walbus Thanks. I'm going to review this https://github.com/spotify/docker-client/pull/266 and then get back to you.

davidxia commented on src/main/java/com/spotify/docker/client/DefaultDockerClient.java:
```diff
@@ -1176,16 +1176,64 @@ private String authRegistryHeader(final AuthRegistryConfig authRegistryConfig)
+   * The result is zero if the strings are _numerically_ equal.
+   * N.B. It does not work if "1.10" is supposed to be equal to "1.10.0".
+   */
+  private int versionCompare(String str1, String str2) {
```
We already have a `versionCompare()` in `DefaultDockerClientTest`. Want to just use that one?

davidxia commented on src/main/java/com/spotify/docker/client/DefaultDockerClient.java:
```diff
@@ -1176,16 +1176,64 @@ private String authRegistryHeader(final AuthRegistryConfig authRegistryConfig)
       String authRegistryJson =
         ObjectMapperProvider.objectMapper().writeValueAsString(authRegistryConfig);
+
+      String apiVersion = version().apiVersion();
```
nit: final

walbus commented:
@davidxia 
I moved the version compare into its own class, and I made the two assignments in authRegistryHeader final.

davidxia commented:
@walbus Hey, sorry to make you do this again. But you'll need to rebase :(

walbus commented:
@davidxia I think I made it work this time :)

davidxia commented:
@walbus awesome thanks!
